File: java/iroha_android/src/main/java/org/hyperledger/iroha/android/nexus/UaidPortfolioResponse.java
package org.hyperledger.iroha.android.nexus;

import java.util.List;
import java.util.Objects;

/** Immutable view over `/v1/accounts/{uaid}/portfolio` responses. */
public final class UaidPortfolioResponse {
  private final String uaid;
  private final UaidPortfolioTotals totals;
  private final List<UaidPortfolioDataspace> dataspaces;

  public UaidPortfolioResponse(
      final String uaid,
      final UaidPortfolioTotals totals,
      final List<UaidPortfolioDataspace> dataspaces) {
    this.uaid = Objects.requireNonNull(uaid, "uaid");
    this.totals = Objects.requireNonNull(totals, "totals");
    this.dataspaces = List.copyOf(Objects.requireNonNull(dataspaces, "dataspaces"));
  }

  public String uaid() {
    return uaid;
  }

  public UaidPortfolioTotals totals() {
    return totals;
  }

  public List<UaidPortfolioDataspace> dataspaces() {
    return dataspaces;
  }

  /** Aggregated account/position totals for a UAID. */
  public static final class UaidPortfolioTotals {
    private final long accounts;
    private final long positions;

    public UaidPortfolioTotals(final long accounts, final long positions) {
      this.accounts = Math.max(0L, accounts);
      this.positions = Math.max(0L, positions);
    }

    public long accounts() {
      return accounts;
    }

    public long positions() {
      return positions;
    }
  }

  /** Dataspace entry within the aggregated portfolio response. */
  public static final class UaidPortfolioDataspace {
    private final long dataspaceId;
    private final String dataspaceAlias;
    private final List<UaidPortfolioAccount> accounts;

    public UaidPortfolioDataspace(
        final long dataspaceId,
        final String dataspaceAlias,
        final List<UaidPortfolioAccount> accounts) {
      this.dataspaceId = dataspaceId;
      this.dataspaceAlias = dataspaceAlias;
      this.accounts = List.copyOf(Objects.requireNonNull(accounts, "accounts"));
    }

    public long dataspaceId() {
      return dataspaceId;
    }

    public String dataspaceAlias() {
      return dataspaceAlias;
    }

    public List<UaidPortfolioAccount> accounts() {
      return accounts;
    }
  }

  /** Account entry underneath a dataspace portfolio record. */
  public static final class UaidPortfolioAccount {
    private final String accountId;
    private final String label;
    private final List<UaidPortfolioAsset> assets;

    public UaidPortfolioAccount(
        final String accountId, final String label, final List<UaidPortfolioAsset> assets) {
      this.accountId = Objects.requireNonNull(accountId, "accountId");
      this.label = label;
      this.assets = List.copyOf(Objects.requireNonNull(assets, "assets"));
    }

    public String accountId() {
      return accountId;
    }

    public String label() {
      return label;
    }

    public List<UaidPortfolioAsset> assets() {
      return assets;
    }
  }

  /** Asset balance entry associated with an account. */
  public static final class UaidPortfolioAsset {
    private final String assetId;
    private final String assetDefinitionId;
    private final String quantity;

    public UaidPortfolioAsset(
        final String assetId, final String assetDefinitionId, final String quantity) {
      this.assetId = Objects.requireNonNull(assetId, "assetId");
      this.assetDefinitionId = Objects.requireNonNull(assetDefinitionId, "assetDefinitionId");
      this.quantity = Objects.requireNonNull(quantity, "quantity");
    }

    public String assetId() {
      return assetId;
    }

    public String assetDefinitionId() {
      return assetDefinitionId;
    }

    public String quantity() {
      return quantity;
    }
  }
}

File: kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/nexus/UaidPortfolioResponse.kt
package org.hyperledger.iroha.sdk.nexus

import org.hyperledger.iroha.sdk.numeric.KotodamaQuantity

/** Immutable view over `/v1/accounts/{uaid}/portfolio` responses. */
class UaidPortfolioResponse(
    @JvmField val uaid: String,
    @JvmField val totals: UaidPortfolioTotals,
    dataspaces: List<UaidPortfolioDataspace>,
) {
    init {
        UaidLiteral.canonicalize(uaid, "uaid")
    }

    @JvmField val dataspaces: List<UaidPortfolioDataspace> = dataspaces.toList()

    /** Aggregated account/position totals for a UAID. */
    class UaidPortfolioTotals(
        @JvmField val accounts: Long,
        @JvmField val positions: Long,
    ) {
        init {
            require(accounts >= 0) { "accounts must be non-negative" }
            require(positions >= 0) { "positions must be non-negative" }
        }
    }

    /** Dataspace entry within the aggregated portfolio response. */
    class UaidPortfolioDataspace(
        @JvmField val dataspaceId: Long,
        @JvmField val dataspaceAlias: String?,
        accounts: List<UaidPortfolioAccount>,
    ) {
        @JvmField val accounts: List<UaidPortfolioAccount> = accounts.toList()
    }

    /** Account entry underneath a dataspace portfolio record. */
    class UaidPortfolioAccount(
        @JvmField val accountId: String,
        @JvmField val label: String?,
        assets: List<UaidPortfolioAsset>,
    ) {
        @JvmField val assets: List<UaidPortfolioAsset> = assets.toList()
    }

    /** Asset balance entry associated with an account. */
    class UaidPortfolioAsset(
        @JvmField val assetId: String,
        @JvmField val assetDefinitionId: String,
        @JvmField val quantity: String,
    ) {
        init {
            KotodamaQuantity.parseCanonical(quantity)
        }
    }
}
